### PR TITLE
[CSL-1712] Consider delegation in block-gen

### PR DESCRIPTION
**Pull request by pva701**

### CODE REVIEW

File: lib/src/Pos/Generator/Block/Logic.hs [hunk 1]
```diff
@@ -80,12 +82,14 @@ genBlock eos = do
...
             genPayload slot
             -- We need to know leader's secret key to create a block.
-            leader <-
+            precomputedLeader <-
                 lift $ maybeThrow
                     (BGInternal "no leader")
                     (leaders ^? ix (fromIntegral $ getSlotIndex siSlot))
             secrets <-
                 unInvSecretsMap . view asSecretKeys <$> view blockGenParams
+            cert <- lift $ getPskByIssuer (Right precomputedLeader)
```
Comment (volhovM) [inline lib/src/Pos/Generator/Block/Logic.hs]: Is it alright that transitive delegation is not considered here?

**Comment by pva701** [comment]
@volhovM @gromakovsky I've fixed and force pushed (not tested yet)

File: lib/src/Pos/Generator/Block/Logic.hs [hunk 2]
```diff
@@ -80,12 +82,14 @@ genBlock eos = do
         EpochOrSlot (Right slot@SlotId {..}) -> withCurrentSlot slot $ do
             genPayload slot
             -- We need to know leader's secret key to create a block.
```
Comment (gromakovsky) [inline lib/src/Pos/Generator/Block/Logic.hs]: This comment is obsolete.

File: lib/src/Pos/Generator/Block/Logic.hs [hunk 3]
```diff
@@ -80,12 +82,14 @@ genBlock eos = do
...
-            leader <-
+            precomputedLeader <-
```
Comment (gromakovsky) [inline lib/src/Pos/Generator/Block/Logic.hs]: Well, I don't know why you renamed it (leader is leader even if she delegated), but it's not essential.

File: lib/src/Pos/Generator/Block/Logic.hs [hunk 4]
```diff
@@ -80,12 +82,14 @@ genBlock eos = do
...
                 lift $ maybeThrow
                     (BGInternal "no leader")
                     (leaders ^? ix (fromIntegral $ getSlotIndex siSlot))
             secrets <-
                 unInvSecretsMap . view asSecretKeys <$> view blockGenParams
+            transCert <- lift $ getDlgTransPsk precomputedLeader
+            let leader = maybe precomputedLeader (addressHash . pskDelegatePk . snd) transCert
```
Comment (gromakovsky) [inline lib/src/Pos/Generator/Block/Logic.hs]: It's not leader. It should be called _creator_ or something.

**Comment by gromakovsky** [comment]
And also should be tested.

**Comment by pva701** [comment]
tested, it works

**Pull request merged**